File: src/main/java/com/hst/sitescrapper/model/api/ApiResponse.java
package com.hst.sitescrapper.model.api;

public class ApiResponse {

    private ApiHeader header;
    private Object body;

    public ApiHeader getHeader() {
        return header;
    }

    public void setHeader(ApiHeader header) {
        this.header = header;
    }

    public Object getBody() {
        return body;
    }

    public void setBody(Object body) {
        this.body = body;
    }

    public static ApiResponse of(int code, String message) {
        return of(code, message, null);
    }

    public static ApiResponse of(int code, String message, Object body) {
        ApiResponse apiResponse = new ApiResponse();
        apiResponse.setHeader(ApiHeader.of(code, message));
        apiResponse.setBody(body);

        return apiResponse;
    }


}

File: backend/src/main/kotlin/com/hst/scrapper/global/model/api/ApiResponse.kt
package com.hst.scrapper.global.model.api

import com.hst.scrapper.global.exception.BaseException
import org.springframework.http.HttpStatus

/**
 * @author dlgusrb0808@gmail.com
 */
data class ApiResponse<T>(
    val header: ApiHeader,
    val body: T
) {
    companion object {
        fun from(status: HttpStatus): ApiResponse<Any?> {
            return of(status, null)
        }

        fun <T> of(status: HttpStatus, body: T): ApiResponse<T> {
            return ApiResponse(ApiHeader.from(status), body)
        }

        fun <T> of(statusCode: Int, message: String, body: T): ApiResponse<T> {
            return ApiResponse(ApiHeader(statusCode, message), body)
        }

        fun success(): ApiResponse<Any?> {
            return from(HttpStatus.OK)
        }

        fun <T> success(body: T): ApiResponse<T> {
            return of(HttpStatus.OK, body)
        }

        fun error(exception: BaseException): ApiResponse<Map<String, Any>> {
            return of(exception.status.value(), exception.message, exception.attribute)
        }
    }
}